source/puppet/4.4/reference/static_catalogs.md

```diff
@@ -0,0 +1,124 @@
+
+### Configuring `code_id` and the `static_file_content` endpoint
+
+When requesting the file's content from a static catalog's metadata, the Puppet agent passes the file's path, the catalog's `code_id`, and the requested environment to Puppet Server's new [`static_file_content`][] API endpoint, and the endpoint returns the appropriate version of the file's contents as the `code_content`.
+
+If static catalogs are enabled but none of the relevant Puppet Server settings are configured, the `code_id` parameter defaults to a null value and the `static_file_content` API endpoint always returns the latest content. To populate the `code_id` with a more useful identifier and have the `static_file_content` endpoint return the appropriate version of the file's content, you must specify scripts or commands that provide Puppet with the right information.
+
+Puppet Server locates these commands via the `code-id-command` and `code-content-command` settings in Puppet Server's [`puppetserver.conf`][] file. Puppet Server runs the `code-id-command` each time it compiles a static catalog, and it runs the `code-content-command` each time an agent requests file contents from the `static_file_content` endpoint.
+
+Puppet Server validates the standard output of each of these scripts, and if the output's acceptable, it adds the results to the catalog as their respective parameters' values. This lets you use any versioning or synchronization tools you want, as long as you can write scripts that produce a hash for the `code_id` and code content using the catalog's `code_id` and file's environment.
+
+The `code-id-command` and `code-content-command` scripts can be as simple or complex as necessary. For instance, if files in an environment are simply managed by git, a `code-id-command` script can be nothing more than this, with the environment name passed to the script as the first command-line argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ -z $1 ]]; then
+  echo Expected an environment >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/$1 && git rev-parse HEAD
+```
+
+As long as the script's exit code is zero, Puppet Server uses the script's standard output as the catalog's `code_id`. If the script returns a non-zero exit code, Puppet Server logs an error and returns the error and a 500 response code to the API request.
+
+A `code-content-command` script can also be simple. Puppet Server passes the environment name as the first command-line argument, the `code_id` as the second argument, and the path to the file resource from its `content_uri` as the third argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ $# < 3 ]]; then
+  echo Expected environment, code-id, file-path >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/$1 && git show $2:$3
```

Comment:
It might be good to have someone review these with an eye for security related issues, e.g. should `$1`, etc be quoted?